model_analyzer/result/result_comparator.py

```diff
@@ -0,0 +1,161 @@
+            If None, then equal weighting (normal priority comparison)
+        """
+
+        # Index the non gpu metric types
+        self._non_gpu_type_to_idx = {
+            non_gpu_metric_types[i]: i
+            for i in range(len(non_gpu_metric_types))
+        }
+
+        # Index the gpu metric types
+        self._gpu_type_to_idx = {
+            gpu_metric_types[i]: i
+            for i in range(len(gpu_metric_types))
+        }
+
+        self._metric_priorities = metric_priorities
+        self._comparison_threshold_factor = (comparison_threshold_percent *
+                                             1.0) / 100
+
+        # TODO implement a weighted comparison
+        self._weights = weights
+
+    def __call__(self, result1, result2):
+        """
+        Allows an instance of a ResultComparator
+        to behave like a function
+
+        Parameters
+        ----------
+        result1 : RunResult
+            first result to be compared
+        result2 : RunResult
+            second result to be compared
+        
+        Returns
+        -------
+        int 
+            0 
+                if the results are determined
+                to be the same within a threshold
+            1
+                if result1 > result2
+            -1
+                if result1 < result2
+        """
+
+        # For now, average over perf runs for given model_config
+        avg_gpu_metrics1, avg_non_gpu_metrics1 = \
+            self._average_measurements(result1)
+        avg_gpu_metrics2, avg_non_gpu_metrics2 = \
+            self._average_measurements(result2)
+
+        for priority in self._metric_priorities:
+            if priority in self._gpu_type_to_idx:
+                # Get position in measurements of current priority's value
+                metric_idx = self._gpu_type_to_idx[priority]
+                threshold = self._comparison_threshold_factor * avg_non_gpu_metrics1[
+                    metric_idx]
+                value_diff = avg_non_gpu_metrics1[
+                    metric_idx] - avg_non_gpu_metrics2[metric_idx]
+
+                if value_diff > threshold:
+                    return 1
+                elif value_diff < -threshold:
+                    return -1
+            elif priority in self._gpu_type_to_idx:
+                metric_idx = self._gpu_type_to_idx[priority]
+                threshold = self._comparison_threshold_factor * avg_gpu_metrics1[
+                    metric_idx]
+                value_diff = avg_gpu_metrics1[metric_idx] - avg_gpu_metrics2[
+                    metric_idx]
+
+                if value_diff > threshold:
+                    return 1
+                elif value_diff < -threshold:
+                    return -1
+            else:
+                raise TritonModelAnalyzerException(
+                    f"Category unknown for objective : '{priority}'")
+        return 0
+
+    def _average_measurements(self, result):
```

Comment:
Can we make this more general? For example, we can add a new argument `aggregator`, and rename the function to `_aggregate_measurements`. This new argument is a `callable` that will be called on the list of Records that we have here.